File: ➡️📂Core kernel Yaiwes/GitHub-Copilot-SDK/GitHub-Copilot-SDK/nodejs/test/e2e/rpc_session_state_extras.e2e.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { describe, expect, it } from "vitest";
import type { CopilotSession } from "../../src/index.js";
import { approveAll, CopilotClient, RuntimeConnection } from "../../src/index.js";
import { createSdkTestContext, DEFAULT_GITHUB_TOKEN } from "./harness/sdkTestContext.js";

describe("Session-scoped state extras RPC", async () => {
    const { copilotClient: client, env, openAiEndpoint, workDir } = await createSdkTestContext();

    function createClientWithEnv(
        extraEnv: Record<string, string | undefined>,
        token = DEFAULT_GITHUB_TOKEN
    ): CopilotClient {
        return new CopilotClient({
            workingDirectory: workDir,
            env: {
                ...env,
                ...extraEnv,
            },
            logLevel: "error",
            connection: RuntimeConnection.forStdio({ path: process.env.COPILOT_CLI_PATH }),
            gitHubToken: token,
        });
    }

    function createAuthenticatedClient(token: string): CopilotClient {
        return createClientWithEnv(
            {
                COPILOT_DEBUG_GITHUB_API_URL: env.COPILOT_API_URL,
            },
            token
        );
    }

    async function configureAuthenticatedUser(token: string): Promise<void> {
        await openAiEndpoint.setCopilotUserByToken(token, {
            login: "rpc-session-extras-user",
            copilot_plan: "individual_pro",
            endpoints: {
                api: env.COPILOT_API_URL,
                telemetry: "https://localhost:1/telemetry",
            },
            analytics_tracking_id: "rpc-session-extras-tracking-id",
        });
    }

    async function createSession(): Promise<CopilotSession> {
        return client.createSession({ onPermissionRequest: approveAll });
    }

    async function disconnect(session: CopilotSession | undefined): Promise<void> {
        if (!session) {
            return;
        }
        try {
            await session.disconnect();
        } catch {
            // Best-effort cleanup.
        }
    }

    it("should list models for session", { timeout: 120_000 }, async () => {
        const token = "rpc-session-model-list-token";
        await configureAuthenticatedUser(token);
        const authClient = createAuthenticatedClient(token);
        let session: CopilotSession | undefined;
        try {
            await authClient.start();
            session = await authClient.createSession({
                model: "claude-sonnet-4.5",
                onPermissionRequest: approveAll,
            });

            const result = await session.rpc.model.list();

            expect(Array.isArray(result.list)).toBe(true);
            expect(result.list.length).toBeGreaterThan(0);
            expect(
                result.list.some((model) => JSON.stringify(model).includes("claude-sonnet-4.5"))
            ).toBe(true);
        } finally {
            await disconnect(session);
            try {
                await authClient.stop();
            } catch {
                // Best-effort cleanup.
            }
        }
    });

    it("should report session activity when idle", { timeout: 120_000 }, async () => {
        const session = await createSession();
        try {
            const activity = await session.rpc.metadata.activity();

            expect(activity.hasActiveWork).toBe(false);
            expect(activity.abortable).toBe(false);
        } finally {
            await session.disconnect();
        }
    });

    it("should add byok provider and model at runtime", { timeout: 120_000 }, async () => {
        const session = await createSession();
        try {
            const providerName = `sdk-runtime-provider-${Date.now()}-${Math.random().toString(36).slice(2)}`;
            const modelId = "sdk-runtime-model";
            const selectionId = `${providerName}/${modelId}`;

            const added = await session.rpc.provider.add({
                providers: [
                    {
                        name: providerName,
                        type: "openai",
                        wireApi: "completions",
                        baseUrl: "https://api.example.test/v1",
                        apiKey: "runtime-provider-secret",
                        headers: { "X-SDK-Provider": "runtime" },
                    },
                ],
                models: [
                    {
                        provider: providerName,
                        id: modelId,
                        name: "SDK Runtime Model",
                        modelId: "claude-sonnet-4.5",
                        wireModel: "wire-sdk-runtime-model",
                        maxContextWindowTokens: 4096,
                        maxPromptTokens: 3072,
                        maxOutputTokens: 1024,
                        capabilities: {
                            limits: {
                                maxContextWindowTokens: 4096,
                                maxPromptTokens: 3072,
                                maxOutputTokens: 1024,
                            },
                            supports: {
                                reasoningEffort: false,
                                vision: false,
                            },
                        },
                    },
                ],
            });

            expect(added.models).toHaveLength(1);
            expect(JSON.stringify(added.models[0])).toContain(selectionId);
            expect(JSON.stringify(added.models[0])).toContain("SDK Runtime Model");

            const listed = await session.rpc.model.list();
            expect(listed.list.some((model) => JSON.stringify(model).includes(selectionId))).toBe(
                true
            );

            const switched = await session.rpc.model.switchTo({ modelId: selectionId });
            expect(switched.modelId).toBe(selectionId);
            expect((await session.rpc.model.getCurrent()).modelId).toBe(selectionId);
        } finally {
            await session.disconnect();
        }
    });

    it(
        "should return empty completions when host does not provide them",
        { timeout: 120_000 },
        async () => {
            const session = await createSession();
            try {
                const triggers = await session.rpc.completions.getTriggerCharacters();
                expect(triggers.triggerCharacters).toEqual([]);

                const completions = await session.rpc.completions.request({
                    text: "Use @",
                    offset: 5,
                });
                expect(completions.items).toEqual([]);
            } finally {
                await session.disconnect();
            }
        }
    );

    it("should report visibility as unsynced for local session", { timeout: 120_000 }, async () => {
        const session = await createSession();
        try {
            const initial = await session.rpc.visibility.get();
            expect(initial.synced).toBe(false);
            expect(initial.status).toBeUndefined();
            expect(initial.shareUrl).toBeUndefined();

            const set = await session.rpc.visibility.set({ status: "repo" });
            expect(set.synced).toBe(false);
            expect(set.status).toBeUndefined();
            expect(set.shareUrl).toBeUndefined();
        } finally {
            await session.disconnect();
        }
    });

    it("should get and set allowall permissions", { timeout: 120_000 }, async () => {
        const session = await createSession();
        try {
            const initial = await session.rpc.permissions.getMode();
            expect(initial.mode).toBe("manual");

            const enable = await session.rpc.permissions.setMode({ mode: "allow-all" });
            expect(enable.success).toBe(true);
            expect(enable.mode).toBe("allow-all");
            expect((await session.rpc.permissions.getMode()).mode).toBe("allow-all");

            const disable = await session.rpc.permissions.setMode({ mode: "manual" });
            expect(disable.success).toBe(true);
            expect(disable.mode).toBe("manual");
            expect((await session.rpc.permissions.getMode()).mode).toBe("manual");
        } finally {
            try {
                await session.rpc.permissions.setMode({ mode: "manual" });
            } catch {
                // Best-effort reset.
            }
            await session.disconnect();
        }
    });

    it(
        "should get context attribution and heaviest messages after turn",
        { timeout: 120_000 },
        async () => {
            const session = await createSession();
            try {
                const answer = await session.sendAndWait({
                    prompt: "Say CONTEXT_METADATA_OK exactly.",
                });
                expect(answer?.data.content ?? "").toContain("CONTEXT_METADATA_OK");

                const attribution = await session.rpc.metadata.getContextAttribution();
                expect(attribution.contextAttribution).not.toBeNull();
                const contextAttribution = attribution.contextAttribution!;
                expect(contextAttribution.totalTokens).toBeGreaterThan(0);
                expect(contextAttribution.entries.length).toBeGreaterThan(0);
                for (const entry of contextAttribution.entries) {
                    expect(entry.id.trim()).toBeTruthy();
                    expect(entry.kind.trim()).toBeTruthy();
                    expect(entry.label.trim()).toBeTruthy();
                    expect(entry.tokens).toBeGreaterThanOrEqual(0);
                    for (const attribute of entry.attributes ?? []) {
                        expect(attribute.key.trim()).toBeTruthy();
                    }
                }

                const heaviest = await session.rpc.metadata.getContextHeaviestMessages({
                    limit: 2,
                });
                expect(heaviest.totalTokens).toBeGreaterThan(0);
                expect(heaviest.messages.length).toBeLessThanOrEqual(2);
                for (const message of heaviest.messages) {
                    expect(message.id.trim()).toBeTruthy();
                    expect(message.tokens).toBeGreaterThanOrEqual(0);
                }
            } finally {
                await session.disconnect();
            }
        }
    );

    it("should update and clear live subagent settings", { timeout: 120_000 }, async () => {
        const session = await createSession();
        try {
            await expect(
                session.rpc.tools.updateSubagentSettings({
                    subagents: {
                        "general-purpose": {
                            model: "claude-haiku-4.5",
                            effortLevel: "low",
                            contextTier: "default",
                        },
                    },
                })
            ).resolves.toBeDefined();

            await expect(
                session.rpc.tools.updateSubagentSettings({
                    subagents: null,
                })
            ).resolves.toBeDefined();
        } finally {
            await session.disconnect();
        }
    });

    it("should read empty sql todos for fresh session", { timeout: 120_000 }, async () => {
        const session = await createSession();
        try {
            const result = await session.rpc.plan.readSqlTodos();

            expect(result.rows).toBeDefined();
            expect(result.rows).toEqual([]);
        } finally {
            await session.disconnect();
        }
    });

    it("should get telemetry engagement id", { timeout: 120_000 }, async () => {
        const session = await createSession();
        try {
            const result = await session.rpc.telemetry.getEngagementId();

            expect(result).toBeDefined();
        } finally {
            await session.disconnect();
        }
    });

    it("should get current tool metadata after initialization", { timeout: 120_000 }, async () => {
        const session = await createSession();
        try {
            const answer = await session.sendAndWait({ prompt: "What is 2+2?" });
            expect(answer).toBeDefined();

            const result = await session.rpc.tools.getCurrentMetadata();

            expect(result.tools).not.toBeNull();
            expect(result.tools!.length).toBeGreaterThan(0);
            for (const tool of result.tools!) {
                expect(tool.name).toBeTruthy();
                expect(tool.description).toBeDefined();
            }
        } finally {
            await session.disconnect();
        }
    });

    it("should reload session plugins", { timeout: 120_000 }, async () => {
        const session = await createSession();
        try {
            await session.rpc.plugins.reload();

            const plugins = await session.rpc.plugins.list();
            expect(plugins.plugins).toBeDefined();
            for (const plugin of plugins.plugins) {
                expect(plugin.name).toBeTruthy();
            }
        } finally {
            await session.disconnect();
        }
    });
});

File: ➡️📂Core kernel Yaiwes/GitHub-Copilot-SDK/GitHub-Copilot-SDK/go/internal/e2e/rpc_session_state_extras_e2e_test.go
package e2e

import (
	"encoding/json"
	"strings"
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
	"github.com/github/copilot-sdk/go/rpc"
)

func TestRpcSessionStateExtras(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	t.Run("should_list_models_for_session", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		const token = "rpc-session-model-list-token"
		registerProxyUser(t, ctx, token, "rpc-session-extras-user", nil)
		authClient := newAuthenticatedClient(ctx, token)
		defer authClient.ForceStop()

		session := createPortedSession(t, authClient, &copilot.SessionConfig{Model: "claude-sonnet-4.5"})
		defer session.Disconnect()

		result, err := session.RPC.Model.List(t.Context())
		if err != nil {
			t.Fatalf("Model.List failed: %v", err)
		}
		if result.List == nil {
			t.Fatal("Expected non-nil model list")
		}
		if len(result.List) == 0 {
			t.Fatal("Expected non-empty model list")
		}
		found := false
		for _, model := range result.List {
			data, err := json.Marshal(model)
			if err == nil && strings.Contains(string(data), "claude-sonnet-4.5") {
				found = true
				break
			}
		}
		if !found {
			t.Fatalf("Expected model list to include claude-sonnet-4.5, got %+v", result.List)
		}
	})

	t.Run("should_report_session_activity_when_idle", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		session := createPortedSession(t, client, nil)
		defer session.Disconnect()

		activity, err := session.RPC.Metadata.Activity(t.Context())
		if err != nil {
			t.Fatalf("Metadata.Activity failed: %v", err)
		}
		if activity.HasActiveWork {
			t.Fatal("Expected a fresh session to report no active work")
		}
		if activity.Abortable {
			t.Fatal("Expected a fresh session to have nothing abortable")
		}
	})

	t.Run("should_get_and_set_allowall_permissions", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		session := createPortedSession(t, client, nil)
		defer session.Disconnect()
		defer func() {
			_, _ = session.RPC.Permissions.SetMode(t.Context(), &rpc.PermissionsSetModeRequest{Mode: rpc.PermissionModeManual})
		}()

		initial, err := session.RPC.Permissions.GetMode(t.Context())
		if err != nil {
			t.Fatalf("Permissions.GetMode initial failed: %v", err)
		}
		if initial.Mode != rpc.PermissionModeManual {
			t.Fatalf("Expected manual mode on a fresh session, got %q", initial.Mode)
		}

		enable, err := session.RPC.Permissions.SetMode(t.Context(), &rpc.PermissionsSetModeRequest{Mode: rpc.PermissionModeAllowAll})
		if err != nil {
			t.Fatalf("Permissions.SetMode(allow-all) failed: %v", err)
		}
		if !enable.Success || enable.Mode != rpc.PermissionModeAllowAll {
			t.Fatalf("Expected successful allow-all mode change, got %+v", enable)
		}
		afterEnable, err := session.RPC.Permissions.GetMode(t.Context())
		if err != nil {
			t.Fatalf("Permissions.GetMode after allow-all failed: %v", err)
		}
		if afterEnable.Mode != rpc.PermissionModeAllowAll {
			t.Fatalf("Expected allow-all mode, got %q", afterEnable.Mode)
		}

		disable, err := session.RPC.Permissions.SetMode(t.Context(), &rpc.PermissionsSetModeRequest{Mode: rpc.PermissionModeManual})
		if err != nil {
			t.Fatalf("Permissions.SetMode(manual) failed: %v", err)
		}
		if !disable.Success || disable.Mode != rpc.PermissionModeManual {
			t.Fatalf("Expected successful manual mode change, got %+v", disable)
		}
		afterDisable, err := session.RPC.Permissions.GetMode(t.Context())
		if err != nil {
			t.Fatalf("Permissions.GetMode after manual failed: %v", err)
		}
		if afterDisable.Mode != rpc.PermissionModeManual {
			t.Fatalf("Expected manual mode, got %q", afterDisable.Mode)
		}
	})

	t.Run("should_read_empty_sql_todos_for_fresh_session", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		session := createPortedSession(t, client, nil)
		defer session.Disconnect()

		result, err := session.RPC.Plan.ReadSqlTodos(t.Context())
		if err != nil {
			t.Fatalf("Plan.ReadSqlTodos failed: %v", err)
		}
		if result.Rows == nil {
			t.Fatal("Expected non-nil SQL todo rows")
		}
		if len(result.Rows) != 0 {
			t.Fatalf("Expected empty SQL todo rows, got %+v", result.Rows)
		}
	})

	t.Run("should_get_telemetry_engagement_id", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		session := createPortedSession(t, client, nil)
		defer session.Disconnect()

		result, err := session.RPC.Telemetry.GetEngagementId(t.Context())
		if err != nil {
			t.Fatalf("Telemetry.GetEngagementId failed: %v", err)
		}
		if result == nil {
			t.Fatal("Expected non-nil telemetry engagement result")
		}
	})

	t.Run("should_get_current_tool_metadata_after_initialization", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		session := createPortedSession(t, client, nil)
		defer session.Disconnect()

		answer, err := session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "What is 2+2?"})
		if err != nil {
			t.Fatalf("SendAndWait failed: %v", err)
		}
		if answer == nil {
			t.Fatal("Expected a final assistant message")
		}

		result, err := session.RPC.Tools.GetCurrentMetadata(t.Context())
		if err != nil {
			t.Fatalf("Tools.GetCurrentMetadata failed: %v", err)
		}
		if result.Tools == nil {
			t.Fatal("Expected non-nil current tool metadata")
		}
		if len(result.Tools) == 0 {
			t.Fatal("Expected non-empty current tool metadata")
		}
		for _, tool := range result.Tools {
			if strings.TrimSpace(tool.Name) == "" {
				t.Fatalf("Expected non-empty tool name, got %+v", tool)
			}
			if strings.TrimSpace(tool.Description) == "" {
				t.Fatalf("Expected non-empty tool description, got %+v", tool)
			}
		}
	})

	t.Run("should_add_byok_provider_and_model_at_runtime", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		session := createPortedSession(t, client, nil)
		defer session.Disconnect()

		apiKey := "provider-key"
		providerType := rpc.ProviderConfigTypeOpenai
		wireAPI := rpc.ProviderConfigWireAPICompletions
		modelName := "Go Added Model"
		maxPromptTokens := float64(4096)
		result, err := session.RPC.Provider.Add(t.Context(), &rpc.ProviderAddRequest{
			Providers: []rpc.NamedProviderConfig{{
				Name:    "go-e2e-provider",
				Type:    &providerType,
				BaseURL: "https://models.example.test/v1",
				APIKey:  &apiKey,
				Headers: map[string]string{"x-provider": "go"},
				WireAPI: &wireAPI,
			}},
			Models: []rpc.ProviderModelConfig{{
				ID:              "small",
				Provider:        "go-e2e-provider",
				Name:            &modelName,
				MaxPromptTokens: &maxPromptTokens,
			}},
		})
		if err != nil {
			t.Fatalf("Provider.Add failed: %v", err)
		}
		if len(result.Models) != 1 {
			t.Fatalf("Expected one added provider model, got %+v", result.Models)
		}

		selectionID := "go-e2e-provider/small"
		if _, err := session.RPC.Model.SwitchTo(t.Context(), &rpc.ModelSwitchToRequest{ModelID: selectionID}); err != nil {
			t.Fatalf("Model.SwitchTo added model failed: %v", err)
		}
		current, err := session.RPC.Model.GetCurrent(t.Context())
		if err != nil {
			t.Fatalf("Model.GetCurrent after provider add failed: %v", err)
		}
		if current.ModelID == nil || *current.ModelID != selectionID {
			t.Fatalf("Expected current model %q, got %+v", selectionID, current)
		}
	})

	t.Run("should_return_empty_completions_when_host_does_not_provide_them", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		session := createPortedSession(t, client, nil)
		defer session.Disconnect()

		result, err := session.RPC.Completions.Request(t.Context(), &rpc.CompletionsRequestRequest{
			Text:   "Use @ to mention context",
			Offset: 5,
		})
		if err != nil {
			t.Fatalf("Completions.Request failed: %v", err)
		}
		if result.Items == nil {
			t.Fatal("Expected non-nil completion items list")
		}
	})

	t.Run("should_report_visibility_as_unsynced_for_local_session", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		session := createPortedSession(t, client, nil)
		defer session.Disconnect()

		status := rpc.SessionVisibilityStatusUnshared
		set, err := session.RPC.Visibility.Set(t.Context(), &rpc.VisibilitySetRequest{Status: status})
		if err != nil {
			t.Fatalf("Visibility.Set failed: %v", err)
		}
		if set.Synced || set.Status != nil || set.ShareURL != nil {
			t.Fatalf("Expected unsynced visibility set result, got %+v", set)
		}
		get, err := session.RPC.Visibility.Get(t.Context())
		if err != nil {
			t.Fatalf("Visibility.Get failed: %v", err)
		}
		if get.Synced || get.Status != nil || get.ShareURL != nil {
			t.Fatalf("Expected unsynced visibility get result, got %+v", get)
		}
	})

	t.Run("should_get_context_attribution_and_heaviest_messages_after_turn", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		session := createPortedSession(t, client, nil)
		defer session.Disconnect()

		answer, err := session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "Say CONTEXT_METADATA_OK exactly."})
		if err != nil {
			t.Fatalf("SendAndWait failed: %v", err)
		}
		if answer == nil {
			t.Fatal("Expected final assistant message")
		}

		attribution, err := session.RPC.Metadata.GetContextAttribution(t.Context())
		if err != nil {
			t.Fatalf("Metadata.GetContextAttribution failed: %v", err)
		}
		if attribution == nil {
			t.Fatal("Expected attribution result")
		}
		limit := int64(5)
		heaviest, err := session.RPC.Metadata.GetContextHeaviestMessages(t.Context(), &rpc.MetadataContextHeaviestMessagesRequest{Limit: &limit})
		if err != nil {
			t.Fatalf("Metadata.GetContextHeaviestMessages failed: %v", err)
		}
		if heaviest.Messages == nil {
			t.Fatal("Expected non-nil heaviest messages list")
		}
	})

	t.Run("should_update_and_clear_live_subagent_settings", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		session := createPortedSession(t, client, nil)
		defer session.Disconnect()

		contextTier := rpc.SubagentSettingsEntryContextTierLongContext
		model := "gpt-5-mini"
		reasoningEffort := "low"
		update, err := session.RPC.Tools.UpdateSubagentSettings(t.Context(), &rpc.UpdateSubagentSettingsRequest{
			Subagents: &rpc.SubagentSettings{
				DisabledSubagents: []string{"legacy-agent"},
				Agents: map[string]rpc.SubagentSettingsEntry{
					"general-purpose": {
						ContextTier: &contextTier,
						Model:       &model,
						EffortLevel: &reasoningEffort,
					},
				},
			},
		})
		if err != nil {
			t.Fatalf("Tools.UpdateSubagentSettings failed: %v", err)
		}
		if update == nil {
			t.Fatal("Expected update result")
		}

		clear, err := session.RPC.Tools.UpdateSubagentSettings(t.Context(), &rpc.UpdateSubagentSettingsRequest{})
		if err != nil {
			t.Fatalf("Tools.UpdateSubagentSettings clear failed: %v", err)
		}
		if clear == nil {
			t.Fatal("Expected clear result")
		}
	})

	t.Run("should_reload_session_plugins", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		session := createPortedSession(t, client, nil)
		defer session.Disconnect()

		if _, err := session.RPC.Plugins.Reload(t.Context()); err != nil {
			t.Fatalf("Plugins.Reload failed: %v", err)
		}
		plugins, err := session.RPC.Plugins.List(t.Context())
		if err != nil {
			t.Fatalf("Plugins.List failed: %v", err)
		}
		if plugins.Plugins == nil {
			t.Fatal("Expected non-nil session plugin list")
		}
		for _, plugin := range plugins.Plugins {
			if strings.TrimSpace(plugin.Name) == "" {
				t.Fatalf("Expected non-empty plugin name, got %+v", plugin)
			}
		}
	})
}
